projects/dendrites/subnetworks/utils.py

```diff
@@ -0,0 +1,383 @@
+# flake8: noqa
+
+import math
+import matplotlib.pyplot as plt
+import numpy as np
+import torch
+
+from sklearn import datasets
+from sklearn.preprocessing import MinMaxScaler
+from sklearn.model_selection import train_test_split
+from sklearn.metrics import accuracy_score
+from torch import nn
+
+from nupic.research.frameworks.dendrites import DendriteSegments
+
+__all__ = [
+    "split_data",
+    "apply_target_transform",
+    "add_task_embedding",
+    "prep_dataset",
+    "load_datasets",
+    "evaluate",
+    "train",
+    "run_experiment",
+    "plot_activations",
+    "StandardMLP",
+    "DendriticMLP",
+]
+
+# ------ Dataset
+def split_data(X, Y, random_state=None):
+    """Split with random selection
+    Return two tuples, one with X and another with Ys"""
+    X_train, X_test, Y_train, Y_test = train_test_split(
+        X, Y, test_size=0.2, random_state=random_state, stratify=Y)
+    return (X_train, X_test), (Y_train, Y_test)
+
+def apply_target_transform(Y, offset=0):
+    """Transform label for continual learning problems"""
+    return [label + offset for label in Y]
+
+
+def add_task_embedding(X, task_num=1, units=2):
+    """Adds a task embedding to the input"""
+    assert task_num < units, \
+        "Task num are 0-indexed should be equal to max units-1"
+
+    num_rows = X.shape[0]
+    embedding = np.zeros((num_rows, units))
+    embedding[:, task_num] = 1
+    return np.hstack([X, embedding])
+
+
+def prep_dataset(datasets, embed_task=False, random_state=None,
+                 target_transform=False, target_transform_offset=3):
+    """Accepts an array of tuples (X, Y, name) or a single tuple
+    Return an array of tuples (Xs, Ys, name) or a single tuple
+    where Xs and Ys are also a tuple containing data for training and test"""
+
+    if type(datasets) != list:
+        datasets = [datasets]
+
+    ready_datasets = []
+    for idx, (X, Y, dataset_name) in enumerate(datasets):
+        if target_transform:
+            Y = apply_target_transform(Y, target_transform_offset * idx)
+        if embed_task:
+            X = add_task_embedding(X, task_num=idx)
+        Xs, Ys = split_data(X, Y, random_state=random_state)
+        ready_datasets.append((Xs, Ys, dataset_name))
+
+    if len(ready_datasets) == 1:
+        return ready_datasets[0]
+    return ready_datasets
+
+
+def load_datasets():
+    """Load Iris, Wine and Mixed datasets"""
```

Comment:
Can you please describe exactly what is returned by this function so that others will know how to call it appropriately?

Reply:
Done